File: packages/adapters/src/venues/apex.ts
import {
  type AssetClass,
  classifyNonCrypto,
  type FundingEvent,
  type FundingSnapshot,
} from "@ai-rates/core";
import { CircuitOpenError, type HttpClient } from "../http";
import { marketRef, mul, num, selectRefreshBatch } from "../parse";
import type { SnapshotBatch, VenueAdapter } from "../types";

/**
 * ApeX Omni.
 *
 * REQUESTS: `GET /v3/ticker` answers one symbol at a time (without `symbol` it returns `[]`, and no
 * all-tickers route exists: `/v3/tickers`, `/v3/all-ticker` and `/v3/ticker/all` are 404, `/v3/funding`
 * needs API-key headers). So each cycle is `GET /v3/symbols` once an hour (~750 KB) plus
 * `TICKER_BUDGET` tickers, least recently fetched first. With 125 tradable contracts on 2026-09-13
 * that is 64 requests a cycle and every contract re-read every 2 cycles (~2 minutes), inside the
 * screener's 5-minute window. The limit is "600 requests per 60 secs Per IP"
 * (https://api-docs.omni.apex.exchange/); 150ms spacing puts a cycle's 64 starts in ~10s.
 *
 * WHY ONLY THIS CYCLE'S SLICE IS EMITTED: the collector appends snapshots to `funding_snapshots`
 * with no conflict key, so repeating a ticker read in a later cycle would store it twice. A contract
 * appears in the cycles that read it, and the 2-minute sweep keeps it in `market_latest`. There is
 * nothing for `warmUp` to seed, since the rate is the per-symbol read.
 *
 * FUNDING -- which field. `fundingRate` is the running estimate for the settlement at
 * `nextFundingTime`, so it is `predicted`. The docs only call it "Funding rate" and its neighbour
 * "Predicted funding rate", so this was measured: sampled every minute on 2026-09-13, BTC's
 * `fundingRate` read -0.00001042 at 22:59:31 and `/v3/history-funding` recorded -0.00001020 for the
 * 23:00 settlement; ETH read -0.00000722 and settled -0.00000698. `predictedFundingRate` is NOT
 * the prediction, whatever its name: it read 0.0000125 on BTC, ETH and SPCX every minute, which is
 * the contracts' `fundingInterestRate` of 0.0003 a day over 24 hours -- the interest component
 * alone -- and neither settlement was anywhere near it.
 *
 * FUNDING -- units and period. A fraction for ONE hour, positive = longs pay. "Funding fees will be
 * exchanged between long and short position holders every 1 hour" and "Funding Fees = Position
 * Value * Index Price * Funding Rate" (https://api-docs.omni.apex.exchange/); history rows are
 * exactly 3,600,000ms apart. Against Hyperliquid's 0.0000125/h the same hour BTC read -0.0000103/h:
 * the same order of magnitude (a percent reading would be 100x off, an 8h one 8x), with the
 * opposite sign because ApeX's BTC was marking 4bp under its index (76,761.71 vs 76,794.20).
 *
 * UNITS, checked 2026-09-13: `openInterest` is base units (BTC 1,578 x 76,737 = $121M), `turnover24h`
 * is quote (BTC $565M against `volume24h` 7,337 BTC at ~77k). `nextFundingTime` is ISO-8601.
 *
 * TRADABILITY: `/v3/symbols` has three lists under `contractConfig`. Kept: `perpetualContract` and
 * `stockContract` rows with `enableTrade`, `enableDisplay` and `enableOpenPosition` all true -- 86 of
 * 138 and 39 of 47 on 2026-09-13; the rest are delisted (TON, MKR, IWM ...: all three false) or
 * close-only (IO: tradable but not displayed or openable). `predictionContract` (184 rows, e.g.
 * `Donald_Trump_win_Presidential_Election_2028`) is left out: those are event contracts priced on
 * an outcome, not perpetuals on an asset, and 178 of them are not even displayed.
 *
 * CLASS: `perpetualContract` rows are crypto (their `category` is a sector: L1, DEFI, MEME ...;
 * PAXG has none). `stockContract` rows declare `category` STOCK 28, COMMODITY 6, INDEX 4 or nothing
 * (SOXL). INDEX is passed as index and `marketRef` settles SPY, QQQ, EWY and DRAM as equity; a missing
 * or unknown category goes to `classifyNonCrypto`.
 *
 * QUOTE: `settleAssetId`, USDT on every contract.
 *
 * BASE: `symbol` (`BTC-USDT`) is the venue symbol, because the funding history is keyed by it; the
 * ticker is asked for `crossSymbolName` (`BTCUSDT`). The parser agrees with the declared
 * `baseTokenId` on all 185 perpetual and stock contracts except the four thousand-unit ones
 * (1000PEPE, 1000BONK, 1000SHIB, 1000000MOG), which it reads as PEPE x1000 and so on -- the same
 * reading as every other venue's 1000PEPE, so it is kept.
 */

const VENUE = "apex";
export const APEX_API = "https://omni.apex.exchange/api/v3";
export const SYMBOLS_MAX_AGE_MS = 60 * 60_000;
/** Tickers per cycle; see REQUESTS. */
export const TICKER_BUDGET = 64;
const FUNDING_HOURS = 1;
/** `limit` above 100 is refused with "invalid get page size" (measured with 500). */
const HISTORY_PAGE_SIZE = 100;
const HISTORY_MAX_PAGES = 50;

export interface ApexContract {
  symbol: string;
  crossSymbolName: string;
  baseTokenId: string;
  settleAssetId: string;
  enableTrade: boolean;
  enableDisplay: boolean;
  enableOpenPosition: boolean;
  isPrelaunch?: boolean;
  category?: string;
  displayMaxLeverage?: string;
}

export interface ApexSymbols {
  data: {
    contractConfig: {
      perpetualContract?: ApexContract[];
      stockContract?: ApexContract[];
      predictionContract?: ApexContract[];
    };
  };
}

export interface ApexTicker {
  symbol: string;
  fundingRate?: string;
  predictedFundingRate?: string;
  nextFundingTime?: string;
  markPrice?: string;
  indexPrice?: string;
  openInterest?: string;
  turnover24h?: string;
}

export interface ApexFundingRow {
  symbol: string;
  rate: string;
  price?: string;
  fundingTime: number;
}

/** A tradable contract with the list it came from, which is what decides its class. */
export interface ApexMarket {
  contract: ApexContract;
  list: "perpetual" | "stock";
}

export function apexAssetClass(market: ApexMarket, base: string): AssetClass {
  if (market.list === "perpetual") return "crypto";
  switch (market.contract.category?.trim().toUpperCase()) {
    case "STOCK":
      return "equity";
    case "COMMODITY":
      return "commodity";
    case "INDEX":
      return "index";
    default:
      return classifyNonCrypto(base);
  }
}

const isLive = (c: ApexContract) =>
  c.enableTrade && c.enableDisplay && c.enableOpenPosition && !c.isPrelaunch;

/** Tradable perpetual and stock contracts, by `symbol`. */
export function tradableApexContracts(body: ApexSymbols): Map<string, ApexMarket> {
  const config = body.data.contractConfig;
  const markets: [string, ApexMarket][] = [
    ...(config.perpetualContract ?? [])
      .filter(isLive)
      .map((contract): [string, ApexMarket] => [contract.symbol, { contract, list: "perpetual" }]),
    ...(config.stockContract ?? [])
      .filter(isLive)
      .map((contract): [string, ApexMarket] => [contract.symbol, { contract, list: "stock" }]),
  ];
  return new Map(markets);
}

function ref(market: ApexMarket) {
  const parsed = marketRef(VENUE, market.contract.symbol);
  return marketRef(VENUE, market.contract.symbol, {
    quote: market.contract.settleAssetId,
    assetClass: apexAssetClass(market, parsed.base),
  });
}

export function parseApexTicker(
  market: ApexMarket,
  ticker: ApexTicker | null | undefined,
  now: number,
): FundingSnapshot | null {
  const rate = num(ticker?.fundingRate);
  const markPrice = num(ticker?.markPrice);
  if (!ticker || ticker.symbol !== market.contract.crossSymbolName || rate === null) return null;
  if (markPrice === null) return null;
  const next = ticker.nextFundingTime ? Date.parse(ticker.nextFundingTime) : Number.NaN;
  return {
    ...ref(market),
    observedAt: now,
    rate,
    basisHours: FUNDING_HOURS,
    intervalHours: FUNDING_HOURS,
    nextFundingAt: Number.isFinite(next) ? next : null,
    kind: "predicted",
    markPrice,
    indexPrice: num(ticker.indexPrice),
    openInterestUsd: mul(num(ticker.openInterest), markPrice),
    volume24hUsd: num(ticker.turnover24h),
    maxLeverage: num(market.contract.displayMaxLeverage),
  };
}

/** Hourly settlements within [fromMs, toMs], oldest first. */
export function parseApexFunding(
  rows: readonly ApexFundingRow[],
  market: ApexMarket,
  fromMs: number,
  toMs: number,
): FundingEvent[] {
  const base = ref(market);
  const bySettlement = new Map<number, FundingEvent>();
  for (const row of rows) {
    const settledAt = num(row.fundingTime);
    const rate = num(row.rate);
    if (settledAt === null || rate === null || settledAt < fromMs || settledAt > toMs) continue;
    // `price` is not documented as the mark (the docs' example gives it no description), so it is
    // not stored as one.
    bySettlement.set(settledAt, {
      ...base,
      settledAt,
      rate,
      basisHours: FUNDING_HOURS,
      markPrice: null,
    });
  }
  return [...bySettlement.values()].sort((a, b) => a.settledAt - b.settledAt);
}

export interface ApexAdapterOptions {
  tickerBudget?: number;
}

export function createApexAdapter(options: ApexAdapterOptions = {}): VenueAdapter {
  const budget = options.tickerBudget ?? TICKER_BUDGET;
  let symbols: { fetchedAt: number; bySymbol: Map<string, ApexMarket> } | null = null;
  const lastFetched = new Map<string, { fetchedAt: number }>();

  async function loadSymbols(client: HttpClient, now: number) {
    if (symbols && now - symbols.fetchedAt < SYMBOLS_MAX_AGE_MS) return symbols.bySymbol;
    const body = await client.getJson<ApexSymbols>(`${APEX_API}/symbols`);
    if (!body?.data?.contractConfig) throw new Error(`${VENUE}: unexpected symbols response`);
    symbols = { fetchedAt: now, bySymbol: tradableApexContracts(body) };
    return symbols.bySymbol;
  }

  return {
    venueId: VENUE,
    // 600 requests per 60s per IP.
    minIntervalMs: 150,

    async fetchSnapshots(client, now): Promise<SnapshotBatch> {
      const live = await loadSymbols(client, now);
      for (const symbol of lastFetched.keys()) {
        if (!live.has(symbol)) lastFetched.delete(symbol);
      }
      const batch = selectRefreshBatch([...live.keys()], lastFetched, now, budget, 0);
      for (const symbol of batch) lastFetched.set(symbol, { fetchedAt: now });

      const errors: unknown[] = [];
      // The client spaces request starts, so these queue rather than burst.
      const results = await Promise.all(
        batch.map(async (symbol) => {
          const market = live.get(symbol) as ApexMarket;
          try {
            const body = await client.getJson<{ data?: ApexTicker[] }>(
              `${APEX_API}/ticker?symbol=${encodeURIComponent(market.contract.crossSymbolName)}`,
            );
            return parseApexTicker(market, body?.data?.[0], now);
          } catch (error) {
            errors.push(error);
            return null;
          }
        }),
      );
      const snapshots = results.filter((s): s is FundingSnapshot => s !== null);
      if (snapshots.length === 0 && errors.length > 0) {
        throw errors.find((e) => e instanceof CircuitOpenError) ?? errors[0];
      }
      return { snapshots, settled: [] };
    },

    async fetchFundingHistory(client, venueSymbol, fromMs, toMs) {
      const live = await loadSymbols(client, Date.now());
      const market = live.get(venueSymbol);
      if (!market) return [];
      // Newest first; step `endTimeExclusive` back to the oldest row of each page.
      const rows: ApexFundingRow[] = [];
      let end = toMs + 1;
      for (let page = 0; page < HISTORY_MAX_PAGES && end > fromMs; page++) {
        const url = `${APEX_API}/history-funding?symbol=${encodeURIComponent(venueSymbol)}&limit=${HISTORY_PAGE_SIZE}&beginTimeInclusive=${fromMs}&endTimeExclusive=${end}`;
        const body = await client.getJson<{ data?: { historyFunds?: ApexFundingRow[] | null } }>(
          url,
        );
        const batch = body?.data?.historyFunds ?? [];
        if (!Array.isArray(batch)) throw new Error(`${VENUE}: unexpected history-funding response`);
        rows.push(...batch);
        if (batch.length < HISTORY_PAGE_SIZE) break;
        const oldest = Math.min(...batch.map((r) => Number(r.fundingTime)));
        if (!Number.isFinite(oldest) || oldest >= end) break;
        end = oldest;
      }
      return parseApexFunding(rows, market, fromMs, toMs);
    },
  };
}

export const apexAdapter: VenueAdapter = createApexAdapter();

File: apps/collector-go/internal/adapters/apex/apex.go
// Package apex parses ApeX Omni's perpetual and stock contracts.
//
// Ported from packages/adapters/src/venues/apex.ts and pinned to the same fixtures
// (packages/adapters/__fixtures__/apex) and the same expected values as apex.test.ts, so the Go and
// TypeScript parsers cannot drift apart while both are collecting.
//
// FUNDING -- which field. `fundingRate` is the running estimate for the settlement at
// `nextFundingTime`, so it is `predicted`. The docs only call it "Funding rate" and its neighbour
// "Predicted funding rate", so this was measured: sampled every minute on 2026-09-13, BTC's
// `fundingRate` read -0.00001042 at 22:59:31 and `/v3/history-funding` recorded -0.00001020 for the
// 23:00 settlement; ETH read -0.00000722 and settled -0.00000698. `predictedFundingRate` is NOT
// the prediction, whatever its name: it read 0.0000125 on BTC, ETH and SPCX every minute, which is
// the contracts' `fundingInterestRate` of 0.0003 a day over 24 hours -- the interest component
// alone -- and neither settlement was anywhere near it.
//
// FUNDING -- units and period. A fraction for ONE hour, positive = longs pay. "Funding fees will be
// exchanged between long and short position holders every 1 hour" and "Funding Fees = Position
// Value * Index Price * Funding Rate" (https://api-docs.omni.apex.exchange/); history rows are
// exactly 3,600,000ms apart. Against Hyperliquid's 0.0000125/h the same hour BTC read -0.0000103/h:
// the same order of magnitude (a percent reading would be 100x off, an 8h one 8x), with the
// opposite sign because ApeX's BTC was marking 4bp under its index (76,761.71 vs 76,794.20).
package apex

import (
	"sort"
	"strings"
	"time"

	"github.com/belyjelli/ai-rates/collector/internal/adapters"
	"github.com/belyjelli/ai-rates/collector/internal/core"
)

const VenueID = "apex"

// API is the venue's public REST root.
const API = "https://omni.apex.exchange/api/v3"

// FundingHours: ApeX settles every hour and quotes a 1-hour rate. See the package doc for the
// measurement behind both halves of that.
const FundingHours = 1.0

// Contract is one row of `/v3/symbols`, from any of its three lists.
type Contract struct {
	// Symbol ("BTC-USDT") is the venue symbol, because the funding history is keyed by it.
	Symbol string `json:"symbol"`
	// CrossSymbolName ("BTCUSDT") is what `/v3/ticker` is asked for, and what its row echoes back.
	CrossSymbolName string `json:"crossSymbolName"`
	// BaseTokenID is the venue's declared base. NOT read: the symbol parser agrees with it on all
	// 185 perpetual and stock contracts except the four thousand-unit ones (1000PEPE, 1000BONK,
	// 1000SHIB, 1000000MOG), which it reads as PEPE x1000 and so on -- the same reading as every
	// other venue's 1000PEPE, so that reading is kept.
	BaseTokenID string `json:"baseTokenId"`
	// SettleAssetID is the quote currency, USDT on every contract.
	SettleAssetID      string `json:"settleAssetId"`
	EnableTrade        bool   `json:"enableTrade"`
	EnableDisplay      bool   `json:"enableDisplay"`
	EnableOpenPosition bool   `json:"enableOpenPosition"`
	IsPrelaunch        bool   `json:"isPrelaunch"`
	// Category is a sector on perpetuals (L1, DEFI, MEME ...) and a class on stock contracts
	// (STOCK, COMMODITY, INDEX). A JSON null decodes to "", which reads the same as absent -- both
	// fall through to ClassifyNonCrypto, exactly as the TypeScript optional chain does.
	Category           string       `json:"category"`
	DisplayMaxLeverage adapters.Num `json:"displayMaxLeverage"`
}

// ContractConfig holds the three lists `/v3/symbols` publishes. Slices are plain, so an absent list
// is an empty one, which is what the TypeScript `?? []` gives each of them.
type ContractConfig struct {
	PerpetualContract []Contract `json:"perpetualContract"`
	StockContract     []Contract `json:"stockContract"`
	// PredictionContract is decoded but never collected; see TradableContracts.
	PredictionContract []Contract `json:"predictionContract"`
}

// SymbolsData and SymbolsResponse are pointers all the way down so that a body which lost its
// payload is distinguishable from one listing nothing, which is what the TypeScript
// `!body?.data?.contractConfig` guard turns into a thrown error.
type SymbolsData struct {
	ContractConfig *ContractConfig `json:"contractConfig"`
}

type SymbolsResponse struct {
	Data *SymbolsData `json:"data"`
}

// Ticker is one row of `/v3/ticker`, which answers a single symbol at a time.
//
// UNITS, checked 2026-09-13: `openInterest` is base units (BTC 1,578 x 76,737 = $121M), `turnover24h`
// is quote (BTC $565M against `volume24h` 7,337 BTC at ~77k). `nextFundingTime` is ISO-8601.
type Ticker struct {
	Symbol string `json:"symbol"`
	// FundingRate is the running estimate for the settlement at NextFundingTime.
	FundingRate adapters.Num `json:"fundingRate"`
	// PredictedFundingRate is NOT the prediction: it is the interest component alone, constant at
	// 0.0000125 across every contract. Decoded so the test can pin that it is never read; see the
	// package doc for the measurement.
	PredictedFundingRate adapters.Num `json:"predictedFundingRate"`
	NextFundingTime      string       `json:"nextFundingTime"`
	MarkPrice            adapters.Num `json:"markPrice"`
	IndexPrice           adapters.Num `json:"indexPrice"`
	// OpenInterest is BASE units, so it is priced at the mark to reach USD.
	OpenInterest adapters.Num `json:"openInterest"`
	// Turnover24h is already quote currency.
	Turnover24h adapters.Num `json:"turnover24h"`
}

// TickerResponse is the `/v3/ticker` body. Asking without a symbol returns an empty array, and so
// does asking for a symbol the venue does not list.
type TickerResponse struct {
	Data []Ticker `json:"data"`
}

// FundingRow is one settled payment from `/v3/history-funding`.
type FundingRow struct {
	Symbol string       `json:"symbol"`
	Rate   adapters.Num `json:"rate"`
	// Price is NOT documented as the mark (the docs' example gives it no description), so it is not
	// stored as one.
	Price adapters.Num `json:"price"`
	// FundingTime is epoch MILLISECONDS, arriving as a bare JSON number.
	FundingTime adapters.Num `json:"fundingTime"`
}

// HistoryResponse is the `/v3/history-funding` body. A null or absent `historyFunds` decodes to a nil
// slice, which is the TypeScript `?? []`; a value that is not an array fails the decode in the
// client, which is that adapter's "unexpected history-funding response" throw.
type HistoryResponse struct {
	Data struct {
		HistoryFunds []FundingRow `json:"historyFunds"`
	} `json:"data"`
}

// List is which of `/v3/symbols`' lists a contract came from, which is what decides its class.
type List string

const (
	ListPerpetual List = "perpetual"
	ListStock     List = "stock"
)

// Market is a tradable contract with the list it came from.
type Market struct {
	Contract Contract
	List     List
}

// Book is the tradable contracts keyed by `symbol`, in the venue's own order.
//
// Order is kept beside the map because the TypeScript side is a Map, whose insertion order is what
// the rotation sweeps in: ranging a Go map would reshuffle which contracts a cycle reads on every
// run, so the sweep would stop converging.
type Book struct {
	Order   []string
	Markets map[string]Market
}

// isLive is a contract that is actually tradable right now.
func isLive(c Contract) bool {
	return c.EnableTrade && c.EnableDisplay && c.EnableOpenPosition && !c.IsPrelaunch
}

// TradableContracts is the tradable perpetual and stock contracts, by `symbol`.
//
// Kept: `perpetualContract` and `stockContract` rows with `enableTrade`, `enableDisplay` and
// `enableOpenPosition` all true -- 86 of 138 and 39 of 47 on 2026-09-13; the rest are delisted (TON,
// MKR, IWM ...: all three false) or close-only (IO: tradable but not displayed or openable).
// `predictionContract` (184 rows, e.g. `Donald_Trump_win_Presidential_Election_2028`) is left out:
// those are event contracts priced on an outcome, not perpetuals on an asset, and 178 of them are
// not even displayed.
//
// A nil config is an empty book rather than a panic; the adapter rejects a payload-less body before
// it ever gets here, which is where the TypeScript throws.
func TradableContracts(config *ContractConfig) Book {
	book := Book{Markets: map[string]Market{}}
	if config == nil {
		return book
	}
	add := func(contracts []Contract, list List) {
		for _, contract := range contracts {
			if !isLive(contract) {
				continue
			}
			if _, seen := book.Markets[contract.Symbol]; !seen {
				book.Order = append(book.Order, contract.Symbol)
			}
			book.Markets[contract.Symbol] = Market{Contract: contract, List: list}
		}
	}
	add(config.PerpetualContract, ListPerpetual)
	add(config.StockContract, ListStock)
	return book
}

// AssetClassFor is the class ApeX declares for a contract, from the list it came from and, for stock
// contracts, its `category`.
//
// `perpetualContract` rows are crypto (their `category` is a sector: L1, DEFI, MEME ...; PAXG has
// none). `stockContract` rows declare `category` STOCK 28, COMMODITY 6, INDEX 4 or nothing (SOXL).
// INDEX is passed as index and MarketRefFor settles SPY, QQQ, EWY and DRAM as equity; a missing or
// unknown category goes to ClassifyNonCrypto.
func AssetClassFor(market Market, base string) core.AssetClass {
	if market.List == ListPerpetual {
		return core.ClassCrypto
	}
	switch strings.ToUpper(strings.TrimSpace(market.Contract.Category)) {
	case "STOCK":
		return core.ClassEquity
	case "COMMODITY":
		return core.ClassCommodity
	case "INDEX":
		return core.ClassIndex
	default:
		return core.ClassifyNonCrypto(base)
	}
}

// refFor names one market.
//
// The quote is `settleAssetId`, which the venue states outright, so HasQuote is set: the field is
// supplied, exactly as the TypeScript passes `quote:` on every call.
//
// No dex is supplied at all -- ApeX is a DEX but publishes no HIP-3-style sub-dex id -- so HasDex
// stays false and the parsed value (nil for these symbols) survives. Setting HasDex with a nil Dex
// would be the different claim "this venue declares no dex", which the TypeScript does not make.
func refFor(market Market) core.MarketRef {
	parsed := adapters.MarketRefFor(VenueID, market.Contract.Symbol, adapters.Overrides{})
	quote := market.Contract.SettleAssetID
	class := AssetClassFor(market, parsed.Base)
	return adapters.MarketRefFor(VenueID, market.Contract.Symbol, adapters.Overrides{
		Quote:      &quote,
		HasQuote:   true,
		AssetClass: &class,
	})
}

// ParseTicker normalises one `/v3/ticker` row into a snapshot, or nil where there is nothing to
// emit: no rate, no mark, or a row for a different symbol than the one asked for.
func ParseTicker(market Market, ticker *Ticker, now int64) *core.FundingSnapshot {
	if ticker == nil || ticker.Symbol != market.Contract.CrossSymbolName || !ticker.FundingRate.OK {
		return nil
	}
	if !ticker.MarkPrice.OK {
		return nil
	}

	basisHours := FundingHours
	markPrice := ticker.MarkPrice.Ptr()
	return &core.FundingSnapshot{
		MarketRef:     refFor(market),
		ObservedAt:    now,
		Rate:          ticker.FundingRate.Val,
		BasisHours:    basisHours,
		IntervalHours: &basisHours,
		NextFundingAt: parseNextFundingTime(ticker.NextFundingTime),
		Kind:          core.KindPredicted,
		MarkPrice:     markPrice,
		IndexPrice:    ticker.IndexPrice.Ptr(),
		// `openInterest` is BASE units, so it is priced at the mark: 1,578 BTC is $121M, not $1,578.
		OpenInterestUSD: adapters.Mul(ticker.OpenInterest.Ptr(), markPrice),
		Volume24hUSD:    ticker.Turnover24h.Ptr(),
		MaxLeverage:     market.Contract.DisplayMaxLeverage.Ptr(),
	}
}

// parseNextFundingTime reads the ISO-8601 settlement instant as epoch milliseconds, or nil when the
// venue sent nothing readable. The nil stands in for the TypeScript `Number.isFinite(Date.parse(...))`
// guard, which keeps an unreadable stamp out rather than filing the settlement at the epoch.
func parseNextFundingTime(value string) *int64 {
	if value == "" {
		return nil
	}
	parsed, err := time.Parse(time.RFC3339, value)
	if err != nil {
		return nil
	}
	ms := parsed.UnixMilli()
	return &ms
}

// ParseFunding returns hourly settlements within [fromMs, toMs], oldest first.
func ParseFunding(rows []FundingRow, market Market, fromMs, toMs int64) []core.FundingEvent {
	base := refFor(market)

	// Keyed by settlement so a row repeated across a page boundary lands once, the later read
	// winning, exactly as the TypeScript Map does.
	bySettlement := make(map[int64]core.FundingEvent, len(rows))
	for _, row := range rows {
		if !row.FundingTime.OK || !row.Rate.OK {
			continue
		}
		settledAt := int64(row.FundingTime.Val)
		if settledAt < fromMs || settledAt > toMs {
			continue
		}
		bySettlement[settledAt] = core.FundingEvent{
			MarketRef:  base,
			SettledAt:  settledAt,
			Rate:       row.Rate.Val,
			BasisHours: FundingHours,
			// `price` is not documented as the mark, so it is not stored as one.
			MarkPrice: nil,
		}
	}

	settlements := make([]int64, 0, len(bySettlement))
	for settledAt := range bySettlement {
		settlements = append(settlements, settledAt)
	}
	sort.Slice(settlements, func(i, j int) bool { return settlements[i] < settlements[j] })

	events := make([]core.FundingEvent, 0, len(settlements))
	for _, settledAt := range settlements {
		events = append(events, bySettlement[settledAt])
	}
	return events
}
